Corrections and Response Matrix

JoschD opened the pull request.

JoschD commented on omc3/correction/sequence_evaluation.py:
```diff
@@ -0,0 +1,289 @@
+"""
```
Also new issue: to be checked if it can be improved/made faster by Laurents suggestion in https://github.com/MethodicalAcceleratorDesign/MAD-X/issues/832#issuecomment-662996548

fsoubelet replied:
Let's make this into a new issue and not take care of it for this PR

JoschD commented on omc3/global_correct.py:
```diff
@@ -0,0 +1,192 @@
+                              "calculates the response analytically.",)
+    params.add_parameter(name="optics_params", type=str, nargs="+",
+                         default=CORRECTION_DEFAULTS["optics_params"],
+                         choices=('MUX', 'MUY', 'BBX', 'BBY', 'BETX', 'BETY', 'DX', 'DY', 'NDX', 'Q',
```
Constants

JoschD commented on omc3/global_correct.py:
```diff
@@ -0,0 +1,192 @@
+    opt = _add_hardcoded_paths(opt)
+    iotools.create_dirs(opt.output_dir)
+    accel_inst = manager.get_accelerator(accel_opt)
+    print("global_correct entrypoint")
```
`print` should be `LOG.info()` or `LOG.debug()`

JoschD commented on omc3/global_correct.py:
```diff
@@ -0,0 +1,192 @@
+    return opt
+
+
+OPTICS_PARAMS_CHOICES = ('MUX', 'MUY', 'BETX', 'BETY', 'DX', 'DY', 'NDX',
```
constants. and then possibly already used in the help

JoschD commented on omc3/global_correct.py:
```diff
@@ -0,0 +1,192 @@
+    return opt
+
+
+def _add_hardcoded_paths(opt):
```
Should also be defined with constants.
Also `pathlib.Path()`

JoschD commented on omc3/model/manager.py:
```diff
@@ -20,6 +20,7 @@
 
 
 def _get_params():
+    print("LLLLLLLLLL")
```
what?

JoschD replied:
(i.e. remove debug prints)

JoschD commented on omc3/model/manager.py:
```diff
@@ -29,9 +30,14 @@ def _get_params():
     if not isinstance(opt.accel, str):
         # assume it's the class
         return opt.accel
+    print("ccccc", other_opt, opt)
```
debug print

JoschD commented on omc3/model/manager.py:
```diff
@@ -29,9 +30,14 @@ def _get_params():
         # assume it's the class
         return opt.accel
+    print("ccccc", other_opt, opt)
+    myinst = ACCELS[opt.accel](other_opt)
+    print(opt.accel)
+    print("nnnnnn", myinst.modifiers )
```
debug print

JoschD replied:
debug print

JoschD commented on omc3/model/manager.py:
```diff
@@ -29,9 +30,14 @@ def _get_params():
 @entrypoint(_get_params())
 def get_accelerator(opt, other_opt):
     """ Returns accelerator instance. """
+    print("bbbb", other_opt)
```
debug print

JoschD commented on omc3/model/manager.py:
```diff
@@ -41,4 +47,5 @@ def get_parsed_opt(opt, other_opt):
     accel = ACCELS[opt.accel]
     parser = EntryPoint(accel.get_parameters(), strict=True)
     accel_opt = parser.parse(other_opt)
+    print("ffff", accel_opt)
```
debug print

JoschD commented on omc3/model_creator.py:
```diff
@@ -61,7 +61,11 @@ def create_instance_and_model(opt, accel_opt):
         logging.basicConfig(level=numeric_level) # warning level to stderr
 
     create_dirs(opt.outputdir)
+    print("aaaa111", accel_opt)
```
debug print

JoschD commented on omc3/model_creator.py:
```diff
@@ -61,7 +61,11 @@ def create_instance_and_model(opt, accel_opt):
+    print("aaaa111", accel_opt)
+
     accel_inst = manager.get_accelerator(accel_opt)
+    print("aaaa", accel_opt)
+    print("dddddddd",     accel_inst.modifiers)
```
debug print

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+"""
```
Could use a description. And should probably be in `scripts`

lmalina replied:
Should not be in here at all...it is in separate branch for separate issue: https://github.com/pylhc/omc3/commit/3f05be47a88ebbedeb4f0845986696af4f376e5e

fsoubelet replied:
@lmalina do you confirm we can delete the file from this branch?

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+        df.drop(columns=[f"STATBET{plane}", f"SYSBET{plane}", "CORR_ALFABETA",
+                         f"STATALF{plane}", f"SYSALF{plane}"], inplace=True)
+    else:
+        df[f"{ERR}BET{plane}"] = df_err_sum(df, f"{ERR}BET{plane}", f"STDBET{plane}")
```
The output should use constants

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def convert_old_phase(opt, plane, old_file_name="phase", new_file_name=PHASE_NAME):
+    """
```
proper description. Old -> new

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def convert_old_total_phase(opt, plane, old_file_name="phasetot", new_file_name=TOTAL_PHASE_NAME):
+    """
```
proper description. Old -> new

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def convert_old_closed_orbit(opt, plane, old_file_name="CO", new_file_name=ORBIT_NAME):
+    """
```
proper description. Old -> new

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def convert_old_dispersion(opt, plane, old_file_name="D", new_file_name=DISPERSION_NAME):
+    """
```
proper description. Old -> new

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def convert_old_normalised_dispersion(opt, plane, old_file_name="ND", new_file_name=NORM_DISP_NAME):
+    """
```
proper description. Old -> new

JoschD commented on omc3/old_measurement_converter.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def convert_old_coupling(opt, old_file_name="couple"):
+    """
```
proper description. Old -> new

JoschD commented on omc3/response_creator.py:
```diff
@@ -0,0 +1,87 @@
+:author: Joschua Dillys
+"""
+import os
+import pickle
```
any updates on `feather` or `parquette`? pickle is problematic because it often requires the same packes to be installed (with compatible versions) as when the object was pickled.

fsoubelet replied:
I would merge as is and later on make a decision for this, in alignment with the decision made in the ABP Computing group, so that we're all consistent. Considering we're pickling dictionaries sometimes, we could also look at `hdf`.

JoschD commented on omc3/correction/response_twiss.py:
```diff
@@ -0,0 +1,875 @@
+    https://doi.org/10.1103/PhysRevAccelBeams.20.054801
+
+"""
+import pickle
```
any updates on `feather` or `parquette`? pickle is problematic because it often requires the same packes to be installed (with compatible versions) as when the object was pickled.

JoschD commented on omc3/correction/sequence_evaluation.py:
```diff
@@ -0,0 +1,289 @@
+"""
+import multiprocessing
+import os
+import pickle
```
any updates on `feather` or `parquette`? pickle is problematic because it often requires the same packes to be installed (with compatible versions) as when the object was pickled.

JoschD commented on omc3/response_creator.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    LOG.info("Creating response.")
+    accel_inst = manager.get_accelerator(other_opt)
+    print("aaaa", other_opt)
+    print("dddddddd",     accel_inst.modifiers)
+
+ #   accel_inst.modifiers = "model/opticsfile.18"
```
debug print

JoschD commented on tests/inputs/models/inj_beam1/macros/general.macros.madx:
```diff
@@ -0,0 +1,226 @@
+/*
```
shouldn't this be somewhere in the repo already?

fsoubelet replied:
It is in `omc3.model.madx_macros`, should be taken from there.

lmalina reviewed (commented):
The code is still rather similar to Beta-Beat.src in the parts corresponding to response matrix calculation, the correction part is much more modified. I am not sure, it is very useful if I review the code, since I made the modifications (even thought a long time ago).  I think the following is esential:

- Response matrix calculation is thoroughly looked at also by someone else than @JoschD and myself. 
- The correction calculation is verified by @JoschD.

I did not find the test of correction convergence, we discussed, i.e. checking that second iteration is closer than the first one.

EirikJaccheri commented:
> The code is still rather similar to Beta-Beat.src in the parts corresponding to response matrix calculation, the correction part is much more modified. I am not sure, it is very useful if I review the code, since I made the modifications (even thought a long time ago). I think the following is esential:
> 
>     * Response matrix calculation is thoroughly looked at also by someone else than @JoschD and myself.
> 
>     * The correction calculation is verified by @JoschD.
> 
> 
> I did not find the test of correction convergence, we discussed, i.e. checking that second iteration is closer than the first one.

I cant remember discussing iteration convergence, so i have indeed not implemented the test. Is this necessary?

lmalina commented:
> > The code is still rather similar to Beta-Beat.src in the parts corresponding to response matrix calculation, the correction part is much more modified. I am not sure, it is very useful if I review the code, since I made the modifications (even thought a long time ago). I think the following is esential:
> > ```
> > * Response matrix calculation is thoroughly looked at also by someone else than @JoschD and myself.
> > 
> > * The correction calculation is verified by @JoschD.
> > ```
> > 
> > 
> > I did not find the test of correction convergence, we discussed, i.e. checking that second iteration is closer than the first one.
> 
> I cant remember discussing iteration convergence, so i have indeed not implemented the test. Is this necessary?

I would say it not strictly necessary now, but eventually we should have such a test.

fsoubelet commented:
Needs a merge of master too. I will review this weekend, could be good to take care of docstrings and type hinting @EirikJaccheri

fsoubelet commented on omc3/correction/optics_class.py:
```diff
@@ -0,0 +1,100 @@
+
+# if __name__ == '__main__':
+#     import tfs
+#     coupling_from_r_matrix(tfs.read("/afs/cern.ch/work/l/lmalina/2019-09-16/LHCB1/Results/Getllm_for_coupling/twiss_cor.dat"))
```
I took care of this in 1edcbe8, resolving.

fsoubelet commented:
So far, I've moved the coupling RDTs calculations to use `optics_functions`, removed `old_measurement_converter`, cleaned up a lot of the new code and type hinted as much as I could. 

I have to say this has quite a BetaBeat.src-y feel, but it works and CI is fine (appart from codeclimate's complaints).

It would be good if the people who wrote some of that code could take care of the important module documentation / docstrings (@JoschD, @lmalina ?).

fsoubelet commented:
Proper use of constants throughout the PR now, I think once important (module) docstrings are added we can go ahead and merge.

JoschD commented:
TODO (reminder for me):
Tests:
- [ ] Simplify Test by using append model to measurement function from code
- [ ] Unit-test for append model function

Analytical Response: 
- [ ] bv-flag should come from accelerator class 
- [ ] direction-sign does not need to change for all responses
- [ ] Test Beam 2

tpersson commented:
When could we merge this PR? It has been going on for some time now...

JoschD commented:
- No need to change signs for response matrix, as MAD-X should take care of this when writing the twiss (added warning, that it will not work with Beam 4).
- Added fake_measurement creator, to create fake measurements from twiss-models  (+ unit tests)
- Added omc3.corrections.model_diff to calculate optics-differences between twiss-model files (+unit tests)
- Simplified Eiriks tests, separated into different files
- Added unit test for some correction functionality
- Changed varmap and fullresponse output format to hdf5, added read/write functions to be used. These will make a change in format easy for the future, if required. (+ unit tests)
- Tried to change MU to PHASE where it made sense
- Updated doc

fsoubelet commented on omc3/correction/handler.py:
```diff
@@ -0,0 +1,325 @@
+
+            if opt.update_response:
+                LOG.debug("Updating response.")
+                # please look away for the next two lines.
```
:(

fsoubelet commented on tests/unit/test_global_correction.py:
```diff
@@ -0,0 +1,79 @@
+
+@pytest.mark.basic
+@pytest.mark.parametrize('orientation', ('skew', 'normal'))
+def test_read_measurement_data(tmp_path, model_inj_beam1, orientation):
```
Same here about `model_inj_beam1`

fsoubelet commented on setup.py:
```diff
@@ -28,7 +28,9 @@
     "generic-parser>=1.0.6",
     "sdds>=0.1.3",
     "h5py>=2.9.0",
+    "tables>=3.6.0",
```
The `pandas` doc [here](https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html) say the minimum requirement on `PyTables` is `3.5.1`, maybe we use that version instead to ease things? Is there a reason we need `>=3.6.0`?

JoschD replied:
3.5.1 has no wheels for windows or macos >py3.7

fsoubelet reviewed (changes requested):
Just a few annoying points / questions from me :)

JoschD commented on omc3/global_correction.py:
```diff
@@ -0,0 +1,356 @@
+
+
+# Define functions here, to new optics params
+def _get_default_values() -> Dict[str, Dict[str, float]]:
```
They are more or less empirical values. I don't think there is a good "why". What do you think @lmalina @mihofer @tpersson @rogeliotomas

tpersson replied:
Yeah the default values are more based on experience. I would adjust the weights to 1 for f1001R and f1010R. I would also give weight to normalized dispersion, when it comes to betx and bety they should also have some default values in case they are only taken from the k-mod.

fsoubelet replied:
@JoschD I think these changes are all we need before merging.

fsoubelet commented on tests/accuracy/test_response_creator.py:
```diff
@@ -18,6 +18,10 @@
 @pytest.mark.parametrize('orientation', ('skew', 'normal'))
 @pytest.mark.parametrize('creator', ('madx', 'twiss'))
 def test_reponse_accuracy(tmp_path, model_inj_beam1, orientation, creator):
+    """ Tests the accuracy of a newly generated response agains the saved
```
typo ;)

fsoubelet reviewed (approved):
Just a typo left

tpersson reviewed (dismissed):
Very nice! Just two questions:
Has the updated branch been tested with real data? Also, has it been tested for beam1 and beam2? I am happy to have it merged so if not just go ahead anyway but then we should do that later...

JoschD commented on omc3/global_correction.py:
```diff
@@ -0,0 +1,356 @@
+
+
+
+Possible problems and notes (lmalina, 2020):
+ * error-based weights default? likely - but be careful with low tune errors vs svd cut in pseudoinverse
+ * manual creation of pd.DataFrame varslist, deltas? maybe tunes in tfs_pandas single value or a column?
+ * There should be some summation/renaming for iterations
+ * For two beam correction
+ * The two beams can be treated separately until the calculation of correction
+ * Values missing in the response (i.e. correctors of the other beam) shall be treated as zeros
+ * Missing a part that treats the output from LSA
```
There are also these comments from Lukas, where I am mostly not sure about what exactly they mean. Could be transformed into issues?

tpersson replied:
To me these things are not even issues but rather possible improvements. I think these can just be put in one issue labeled improvement for now.

JoschD commented:
> Very nice! Just two questions:
> Has the updated branch been tested with real data? Also, has it been tested for beam1 and beam2? I am happy to have it merged so if not just go ahead anyway but then we should do that later...

They have only been tested on beam1 and not on real data. 
I skipped beam 2 as the only beam dependence is in https://github.com/pylhc/omc3/blob/457e8f951ef351cb70521057af5c41e8a85d8542/omc3/model/accelerators/lhc.py#L352 but it's actually a good point. All tests are also only LHC.

Real data is a bit more complicated to test for accuracy, maybe that the deltas are getting smaller. Or a simple unit test.

JoschD commented:
Testing other accelerators requires #286 being merged first

JoschD commented:
Left some todos:
- Beam 2 doesn't have varmap files, so these should be created (good test if that works anyway) and then the tests with 'twiss' can run
- No tests for PS or PSB response creation implemented. Should work, but is untested

tpersson commented:
What are the varmap? In order to get this one into the master I would recommend that we skip tests for PS and PSB for now. I would fix the relative path and then we merge!

fsoubelet commented:
For the paths, as we get `Path` objects we could use `.absolute()` where relevant so that used and written versions don't leave room for error. It would also be a small change. Thoughts @JoschD ?

tpersson reviewed (approved):
Excellent! Let's merge!!

mihofer reviewed (approved):
no in-depth check as there are a ton of changes, but looks good .. one thing, one could add a section to the omc wiki á la the omc3 intro to illustrate the workflow

fsoubelet reviewed (approved):
Good for me too! I agree with @mihofer that we will need a workflow page in the OMC website for this too. I'll create an issue.
